File: chatbot/includes/SessionManager.php
<?php
/**
 * SessionManager Class
 * Converted from Python session_manager.py to PHP
 * Handles session storage and management
 */

class SessionManager {
    private $sessionTimeout;
    private $storageFile;
    
    /**
     * Initialize session manager
     * 
     * @param int $sessionTimeout Session timeout in seconds (default: 2 hours)
     */
    public function __construct($sessionTimeout = 7200) {
        $this->sessionTimeout = $sessionTimeout;
        $this->storageFile = __DIR__ . '/sessions.json';
        $this->cleanupOldSessions();
    }
    
    /**
     * Load sessions from storage file
     * 
     * @return array Sessions data
     */
    private function loadSessions() {
        if (!file_exists($this->storageFile)) {
            return [];
        }
        
        $data = file_get_contents($this->storageFile);
        if (!$data) {
            return [];
        }
        
        $sessions = json_decode($data, true);
        return is_array($sessions) ? $sessions : [];
    }
    
    /**
     * Save sessions to storage file
     * 
     * @param array $sessions Sessions data to save
     */
    private function saveSessions($sessions) {
        file_put_contents($this->storageFile, json_encode($sessions, JSON_PRETTY_PRINT));
    }
    
    /**
     * Remove expired sessions from memory
     */
    public function cleanupOldSessions() {
        $sessions = $this->loadSessions();
        $currentTime = time();
        $expiredCount = 0;
        
        foreach ($sessions as $sessionId => $data) {
            $lastActivity = $data['last_activity'] ?? 0;
            if (($currentTime - $lastActivity) > $this->sessionTimeout) {
                unset($sessions[$sessionId]);
                $expiredCount++;
            }
        }
        
        if ($expiredCount > 0) {
            $this->saveSessions($sessions);
            error_log("Cleaned up {$expiredCount} expired sessions");
        }
    }
    
    /**
     * Get or create a session
     * 
     * @param string $sessionId Unique session identifier
     * @return array Session data
     */
    public function getSession($sessionId) {
        $sessions = $this->loadSessions();
        
        if (!isset($sessions[$sessionId])) {
            $sessions[$sessionId] = [
                'title' => 'Sir',
                'last_activity' => time()
            ];
            $this->saveSessions($sessions);
        } else {
            // Update last activity
            $sessions[$sessionId]['last_activity'] = time();
            $this->saveSessions($sessions);
        }
        
        return $sessions[$sessionId];
    }
    
    /**
     * Update session data
     * 
     * @param string $sessionId Session identifier
     * @param array $updates Data to update
     */
    public function updateSession($sessionId, $updates) {
        $sessions = $this->loadSessions();
        
        if (!isset($sessions[$sessionId])) {
            $sessions[$sessionId] = [
                'title' => 'Sir',
                'last_activity' => time()
            ];
        }
        
        foreach ($updates as $key => $value) {
            $sessions[$sessionId][$key] = $value;
        }
        
        $sessions[$sessionId]['last_activity'] = time();
        $this->saveSessions($sessions);
    }
    
    /**
     * Delete a specific session
     * 
     * @param string $sessionId Session identifier to delete
     * @return bool True if deleted, false if not found
     */
    public function deleteSession($sessionId) {
        $sessions = $this->loadSessions();
        
        if (isset($sessions[$sessionId])) {
            unset($sessions[$sessionId]);
            $this->saveSessions($sessions);
            return true;
        }
        
        return false;
    }
    
    /**
     * Get the current number of active sessions
     * 
     * @return int Number of active sessions
     */
    public function getSessionCount() {
        $sessions = $this->loadSessions();
        return count($sessions);
    }
    
    /**
     * Get all sessions (for debugging)
     * 
     * @return array All sessions
     */
    public function getAllSessions() {
        return $this->loadSessions();
    }
}
?>

File: chatbot/session_manager.py
# import time
# import threading


# class SessionManager:
#     """Thread-safe session storage with automatic cleanup"""
    
#     def __init__(self, cleanup_interval=3600, session_timeout=7200):
#         """
#         Initialize session manager with cleanup parameters
        
#         Args:
#             cleanup_interval (int): Seconds between cleanup runs (default: 1 hour)
#             session_timeout (int): Session timeout in seconds (default: 2 hours)
#         """
#         self.sessions = {}
#         self.session_lock = threading.RLock()
#         self.cleanup_interval = cleanup_interval
#         self.session_timeout = session_timeout
#         self.start_cleanup_thread()
    
#     def start_cleanup_thread(self):
#         """Start the background cleanup thread"""
#         def cleanup():
#             while True:
#                 time.sleep(self.cleanup_interval)
#                 self.cleanup_old_sessions()
        
#         thread = threading.Thread(target=cleanup, daemon=True)
#         thread.start()
    
#     def cleanup_old_sessions(self):
#         """Remove expired sessions from memory"""
#         current_time = time.time()
#         with self.session_lock:
#             expired_sessions = [
#                 session_id for session_id, data in self.sessions.items()
#                 if current_time - data.get('last_activity', 0) > self.session_timeout
#             ]
#             for session_id in expired_sessions:
#                 del self.sessions[session_id]
#             if expired_sessions:
#                 print(f"Cleaned up {len(expired_sessions)} expired sessions")
    
#     def get_session(self, session_id):
#         """
#         Get or create a session
        
#         Args:
#             session_id (str): Unique session identifier
            
#         Returns:
#             dict: Session data
#         """
#         with self.session_lock:
#             if session_id not in self.sessions:
#                 self.sessions[session_id] = {
#                     'title': 'Sir',
#                     'last_activity': time.time()
#                 }
#             else:
#                 self.sessions[session_id]['last_activity'] = time.time()
#             return self.sessions[session_id]
    
#     def update_session(self, session_id, updates):
#         """
#         Update session data
        
#         Args:
#             session_id (str): Session identifier
#             updates (dict): Data to update
#         """
#         with self.session_lock:
#             session = self.get_session(session_id)
#             session.update(updates)
#             session['last_activity'] = time.time()
    
#     def delete_session(self, session_id):
#         """
#         Delete a specific session
        
#         Args:
#             session_id (str): Session identifier to delete
#         """
#         with self.session_lock:
#             if session_id in self.sessions:
#                 del self.sessions[session_id]
#                 return True
#             return False
    
#     def get_session_count(self):
#         """Get the current number of active sessions"""
#         with self.session_lock:
#             return len(self.sessions)
    
#     def get_all_sessions(self):
#         """Get a copy of all sessions (for debugging)"""
#         with self.session_lock:
#             return dict(self.sessions)